sklearn/preprocessing/_csr_hstack.pyx

```diff
@@ -0,0 +1,103 @@
+# necessary to maintain integrity, determined by the maximum post-concatenation
+# index/indptr value.
+ctypedef fused IND_2_t:
+    cnp.int32_t
+    cnp.int64_t
+
+# TODO Add const keyword to input arrays in Cython 3.X
+cpdef void _csr_hstack(
+    const Py_ssize_t n_blocks,      # Number of matrices to stack
+    const Py_ssize_t n_rows,        # Number of rows (same across all matrices)
+    const cnp.int32_t[:] n_cols,    # Number of columns (one per matrix)
+    IND_1_t[:] indptr_cat,          # Input concatenated array of indptrs
+    IND_1_t[:] indices_cat,         # Input concatenated array of indices
+    DATA_t[:] data_cat,             # Input concatenated array of data
+    IND_2_t[:] indptr,              # Output array to write indptr into
+    IND_2_t[:] indices,             # Output array to write indices into
+    DATA_t[:] data,                 # Output array to write data into
+    ) nogil:
+
+    cdef:
+        Py_ssize_t offset, row_start, row_end, row_sum, idx, jdx, kdx
+        IND_1_t* col_offset
+        IND_1_t** indptr_bound
+        IND_1_t** indices_bound
+        DATA_t** data_bound
+
+    # The bounds will store the locations/indices of the flat concatenated
+    # arrays that correspond to each block. Need to dynamically allocate their
+    # memory.
+    with nogil:
+        col_offset = <IND_1_t*> malloc(n_blocks * sizeof(IND_1_t))
+        indptr_bound = <IND_1_t**> malloc(n_blocks * sizeof(IND_1_t*))
+        indices_bound = <IND_1_t**> malloc(n_blocks * sizeof(IND_1_t*))
+        data_bound = <DATA_t**> malloc(n_blocks * sizeof(DATA_t*))
+
+        # We set the initial pointers/values here and update iteratively
+        col_offset[0] = 0
+        indptr_bound[0] = &indptr_cat[0]
+        indices_bound[0] = &indices_cat[0]
+        data_bound[0] = &data_cat[0]
+
+        # We populate the bounds iteratively
+        for idx in range(1, n_blocks):
+            col_offset[idx] = col_offset[idx - 1] + n_cols[idx - 1]
+            indptr_bound[idx] = indptr_bound[idx - 1] + n_rows + 1
+            indices_bound[idx] = indices_bound[idx - 1] + indptr_bound[idx - 1][n_rows]
+            data_bound[idx] = data_bound[idx - 1] + indptr_bound[idx - 1][n_rows]
+
+        # Now we populate the full output matrix
+        indptr[0] = 0
+        row_sum = 0
+
+        # We iterate across rows for convenience
+        for idx in range(n_rows):
+            # For each row, in each matrix we find the valid indices to iterate
+            for jdx in range(n_blocks):
+                row_start = indptr_bound[jdx][idx]
+                row_end = indptr_bound[jdx][idx+1]
+
+                # We account for the offset due to horizontal concatenation
+                offset = col_offset[jdx]
+
+                # We iterate over the valid indices, updating the indices and
+                # data
+                for kdx in range(row_start, row_end):
+                    indices[row_sum + kdx] = indices_bound[jdx][kdx] + offset
```

Comment:
This line is generating a segmentation fault. Currently in the process of debugging, but if anyone has any ideas please let me know :)

Reply:
Can you post a code snippet that segfaulted?